### PR TITLE
[builtins] Update builtin xrootd to 5.3.0 and move it to proper ROOT builtins

### PR DESCRIPTION
**Pull request by oshadura**

### CODE REVIEW

**Comment by linev** [comment]
I am facing problem to build xrootd on my platform - latest OpenSUSE Tumblweed.
Maybe latest 4.x XROOTD can solve the problem - I will try

**Comment by oshadura** [comment]
> I am facing problem to build xrootd on my platform - latest OpenSUSE Tumblweed.
> Maybe latest 4.x XROOTD can solve the problem - I will try

@linev can you post please what do you as an error?

**Comment by linev** [comment]
> can you post please what do you see as an error?

It is exactly same error as with 4.12.3, which is mentioned in https://github.com/root-project/root/issues/7292

File: builtins/xrootd/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  # Guess under which directory XRootD will install its libraires
+  set(XROOTD_LIBDIR "lib")
+  if(CMAKE_SYSTEM_NAME STREQUAL "Linux" AND ${CMAKE_SIZEOF_VOID_P} EQUAL 8
+     AND NOT CMAKE_CROSSCOMPILING AND NOT EXISTS "/etc/debian_version")
+    set(XROOTD_LIBDIR "lib64")
+  endif()
+
+  set(XROOTD_LIBRARIES ${XROOTD_ROOTDIR}/${XROOTD_LIBDIR}/libXrdUtils${CMAKE_SHARED_LIBRARY_SUFFIX}
+                       ${XROOTD_ROOTDIR}/${XROOTD_LIBDIR}/libXrdClient${CMAKE_SHARED_LIBRARY_SUFFIX}
```
Comment (linev) [inline builtins/xrootd/CMakeLists.txt:26]: @oshadura 
libXrdClient no longer provided in the xrootd 5.x,
If I rermove it here, I can compile ROOT on my system.
Version 5.1 is working for me.
But we have to check that `net/netx` disabled for xrootd 5.x while `net/netx` based on client library

Comment (oshadura) [reply builtins/xrootd/CMakeLists.txt:26]: Thanks a lot for feedback! I will add your suggestion.

>>But we have to check that net/netx disabled for xrootd 5.x while net/netx based on client library
it is already done some time ago.

Comment (linev) [reply builtins/xrootd/CMakeLists.txt:26]: > it is already done some time ago.

Unfortunately not - if I specify `-Dnetx=ON` ROOT will try to build it without checking of XROOTD version.
If you want, I can add these fixes to your PR

Comment (oshadura) [reply builtins/xrootd/CMakeLists.txt]: hmmm, I thought https://github.com/root-project/root/blob/master/cmake/modules/SearchInstalledSoftware.cmake#L980-L989 will work, let me try to build with `-Dnetx=ON`.

**Comment by oshadura** [comment]
@Axel-Naumann I would like to add this PR to ROOT 6.24. Would it be still possible?

File: builtins/xrootd/CMakeLists.txt (hunk 1 above)
Comment (linev) [reply builtins/xrootd/CMakeLists.txt]: I got problem when I call cmake second time adding `-Dnetx=ON`.
But I have anyway problems with incremental builds now, but this is another issue

Comment (linev) [reply builtins/xrootd/CMakeLists.txt]: And now you move setting of xrootd version number in other cmake file - is it a problem?
I propose, that these checks done directly in `net/netx` and `net/netxng` - at the moment when they build

Comment (oshadura) [reply builtins/xrootd/CMakeLists.txt]: Thanks, let me try to experiment with your suggestion.

**Comment by Axel-Naumann** [comment]
That depends on when it's in a mergable state. Especially those https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/108798/testReport/projectroot.tree.treeplayer/test/gtest_tree_treeplayer_test_treeprocessormt_remotefiles/ and friends seem relevant (IIRC @vepadulano had fixed this in the past?); @eguiraud confirmed that the rdf override / overwrite tests are already fixed.

**Comment by vepadulano** [comment]
> Especially those https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/108798/testReport/projectroot.tree.treeplayer/test/gtest_tree_treeplayer_test_treeprocessormt_remotefiles/ and friends seem relevant

These look similar to something that I addressed in the past, that currently is taken care of with https://github.com/root-project/root/blob/master/cmake/modules/XROOTDApplePostInstall.cmake . I tried to look for the messages from that script in the logs of the build (e.g. "Adjusting LC_RPATH variable of XRootD libraries in ${install_libdir}" ) but there was none so maybe that's weird.

That being said, in this particular case the error is that `libNetxNG.6.25.01.so` doesn't find the reference of `@rpath/libXrdUtils.3.dylib` to load, whereas in the past the issue was that all `libXrD*` libraries couldn't load other xrootd libs in the same install folder. So, possibly, it's the RPATH of `libNetxNG` that should be changed in this case

**Comment by linev** [comment]
https://github.com/root-project/root/issues/7292#issuecomment-790696084

Probably 4.12.8 will be better choice for 6.24 release?

**Comment by oshadura** [comment]
I would prefer xrootd 5 with a lot of new features then xrootd 4.

**Comment by linev** [comment]
But do we have enough time to test ROOT with newer xrootd

**Comment by oshadura** [comment]
@vepadulano I forgot if this problem was reproducible on any Mac? On my Mac:
```
  │  ~/CERN_/root-rntuple/builds │   builtin-xrootd-5.1.0 ?1  ctest -R test-stressIOPlugins-xroot                            ✔ │ 16:54:58 
Test project /Users/oshadura/CERN_sources/root-rntuple/builds
    Start 216: test-stressIOPlugins-xroot
1/1 Test #216: test-stressIOPlugins-xroot .......   Passed    7.80 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   7.92 sec
```

**Comment by oshadura** [comment]
> It is not fully correct.
> `netx` and `netxng` are no longer ROOT compiler options and therefore undefined.
> One should just compile or not compile them based on xrootd version

right, I didn't understand that it is not a build option! Fixed!

**Comment by linev** [comment]
Looks like on all platforms xrootd tests failing

**Review by linev — CHANGES_REQUESTED** [review]
Seems to be, xrootd 5.1 version does not work with current ROOT code.
We have to debug it before merge this PR

**Comment by linev** [comment]
@oshadura Can we make intermediate step with version xrootd 4.12.8. 
We only need to get it on the http://lcgpackages.web.cern.ch/ and update makefile.
Otherwise none of 6.22, 6,24 or master can not be compiled and used on newest Linux distributions - not only OpenSUSE

**Comment by oshadura** [comment]
@linev yes I agree, I am opening separate pr.

**Comment by oshadura** [comment]
FYI it is  https://github.com/root-project/root/pull/7380

**Comment by mrodozov** [comment]
@oshadura 
I have a question concerning xrootd & ROOT and it is does xrootd uses anything from ROOT as libraries/support/dependency ? 
I would've opened an issue but this is already open and it seems ROOT uses xrootd (or it might be using it)
So what happens is that the xrootd works fine (at first), opens a file and then ROOT fails in 

```
Thread 1 "cmsRun" received signal SIGSEGV, Segmentation fault.
0x00007ffff7567807 in TMutex::UnLock() ()
```

**Comment by mrodozov** [comment]
> Seems to be, xrootd 5.1 version does not work with current ROOT code.
> We have to debug it before merge this PR

now I'm reading this, but I have a question and it's since xrootd doesn't depend on ROOT how it doesn't work with any version of it ?  I'm missing something 😃

**Comment by Axel-Naumann** [comment]
Just lost in translation, the accusative got lost: ROOT cannot build with a (builtin) Xrootd 5.1.

**Comment by Axel-Naumann** [comment]
@mrodozov can you open a separate ticket for this, please?

**Comment by linev** [comment]
@Axel-Naumann 

One can build ROOT with xrootd 5, by many tests are failing afterwards.

**Comment by mrodozov** [comment]
@Axel-Naumann a wasn't sure if it has anything to do with ROOT. I'm opening another issue with what I have

> ROOT cannot build with a (builtin) Xrootd 5.1.

got it.

**Comment by oshadura** [comment]
@Axel-Naumann I can reproduce failures on my machine...
Here is quick reproducer:
https://github.com/root-project/root/blob/master/tree/treeplayer/test/treeprocmt/treeprocessormt_remotefiles.cxx#L13-L15
```
root [1] auto fname = "root://eospublic.cern.ch//eos/root-eos/cms_opendata_2012_nanoaod/ZZTo4mu.root";
root [2] auto f = std::unique_ptr<TFile>(TFile::Open(fname));
root [3] f
(std::unique_ptr<TFile, std::default_delete<TFile> > &) std::unique_ptr -> nullptr
root [4]
```
broken xrootd version?

**Comment by oshadura** [comment]
Update: looks like I am missing client library, debugging...

**Comment by Axel-Naumann** [comment]
Yes, or broken config, or who knows what :-) But the latest Xrootd 5 used in production should work. Can you try that one (whatever it is) and if that fails, too, talk to Michal to see what we're doing wrong?

**Comment by Axel-Naumann** [comment]
@phsft-bot build

**Comment by eguiraud** [comment]
@phsft-bot build

**Comment by oshadura** [comment]
now we have an issues with rpath on two mac nodes, 
```
cling::DynamicLibraryManager::loadLibrary(): dlopen(/Users/sftnight/build/workspace/root-pullrequests-build/build/lib/libNetxNG.6.25.01.so, 9): Library not loaded: @rpath/libXrdUtils.3.dylib
  Referenced from: /Users/sftnight/build/workspace/root-pullrequests-build/build/lib/libNetxNG.6.25.01.so
  Reason: image not found
```
  │  ~/CERN_/root/builds │   builtin-xrootd-5.1.0 *6 ?1  otool -l lib/libNetxNG.so | grep RPATH -A2      ✔ │ 18:11:13 
          cmd LC_RPATH
      cmdsize 64
         path /Users/oshadura/CERN_sources/root/builds/lib (offset 12)
--
          cmd LC_RPATH
      cmdsize 88
         path /Users/oshadura/CERN_sources/root/builds/builtins/xrootd/XROOTD-prefix/lib (offset 12)
```
but I can't reproduce it locally on my mac:

**Comment by oshadura** [comment]
@phsft-bot  build!

**Comment by oshadura** [comment]
@bellenot @Axel-Naumann  it think it is ready to be reviewed! Big thanks to @simonmichal!

**Comment by Axel-Naumann** [comment]
@phsft-bot build with flags -Dbuildin_xrootd=Off

**Comment by Axel-Naumann** [comment]
The failures are unrelated.

**Review by Axel-Naumann — COMMENTED** [review]
LGTM, thanks a lot, @oshadura and @simonmichal! We still need @linev 's green light, though - so just commenting to not bypass his review.

**Comment by linev** [comment]
Now ROOT compiles and tests are running.
Last question for @oshadura  - was compilation tested with older 4.x xrootd?

**Comment by Axel-Naumann** [comment]
@phsft-bot build with flags -Dbuiltin_xrootd=Off

**Comment by Axel-Naumann** [comment]
(@linev : that's what I was trying to find out with disabling the builtin xrootd but I had a spelling mistake argh.)

**Comment by linev** [comment]
Looks like compilation was ok - on platforms with installed xrood.

**Review by linev — APPROVED** [review]

**Comment by Axel-Naumann** [comment]
Right, ubuntu 16 built with xrootd 4.8.5. @oshadura assigning to you so you have the pleasure to merge :-)

**Comment by Axel-Naumann** [comment]
FYI there's still the warning:
```
builtins/xrootd/XROOTD-prefix/include/xrootd/XrdCl/XrdClOptional.hh:58:29: warning: unused parameter 'n' [-Wunused-parameter]
```

**Comment by simonmichal** [comment]
> FYI there's still the warning:
> 
> ```
> builtins/xrootd/XROOTD-prefix/include/xrootd/XrdCl/XrdClOptional.hh:58:29: warning: unused parameter 'n' [-Wunused-parameter]
> ```

@Axel-Naumann : I'll fix this in next xrootd release, right now there's not much I can do.

**Comment by Axel-Naumann** [comment]
Understood but we can (and should) use `-isystem` for xrootd includes as we can do even less about this! :-)

**Pull request merged**